### PR TITLE
Initial OAuth2 + API work

### PR DESCRIPTION
**Pull request by nekodex**
Let's start getting the base OAuth2 stuff in so that LAZER™ can start using the endpoints with live data.

Things to note:
* API is very incomplete and the structure is still in flux
* There will be some... hacky things in place until the existing infrastructure has transitioned over
* Third-party access to the API is not yet available

### CODE REVIEW

File: app/Http/Controllers/API/ChatController.php [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+            new PrivateMessageTransformer()
+        );
+    }
+
+    public function postMessage()
+    {
+        $current_user = User::find(Authorizer::getResourceOwnerId());
+
+        if ($current_user->isBanned() || $current_user->isRestricted() || $current_user->isSilenced()) {
+            return $this->error('not authorized', 401);
```
Comment (nanaya) [inline app/Http/Controllers/API/ChatController.php]: ４０３?

File: app/Http/Controllers/API/ChatController.php [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+            case 'channel':
+                $target = Channel::findOrFail(Request::input('channel_id'));
+                break;
+            case 'user':
+                $target = User::findOrFail(Request::input('user_id'));
+                break;
+        }
+
+        if (!$target || !$target->sendMessage($current_user, Request::input('message'))) {
+            return $this->error('not authorized', 401);
```
Comment (nanaya) [inline app/Http/Controllers/API/ChatController.php]: cool story

File: app/Http/Controllers/OAuthController.php [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+use Authorizer;
+use DB;
+
+class OAuthController extends Controller
+{
+    protected $section = 'oauth';
+
+    public function __construct()
+    {
+        // TODO: uh yeah, there's gotta be a better way to do this :v
```
Comment (nanaya) [inline app/Http/Controllers/OAuthController.php]: like an `['except' => ['getAccessToken']]`? :v

File: app/Models/Chat/Channel.php [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        'creation_time',
+    ];
+    public $timestamps = false;
+
+    public function messages()
+    {
+        $this->hasMany(Message::class, 'channel_id', 'channel_id');
+    }
+
+    public function allowedGroups()
```
Comment (nanaya) [inline app/Models/Chat/Channel.php]: there's this getAllowedGroupsAttribute magic

File: app/Models/Chat/Channel.php [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+            case 'spectator':
+            case 'multiplayer':
+            case 'temporary': // this and the comparisons below are needed until bancho is updated to use the new channel types
+                if (substr($this->name, 0, 7) === '#spect_') {
+                    return true;
+                }
+                if (substr($this->name, 0, 4) === '#mp_') {
+                    $match_id = intval(str_replace('#mp_', '', $this->name));
+
+                    return in_array($user->user_id, Match::find($match_id)->currentPlayers(), true);
```
Comment (nanaya) [inline app/Models/Chat/Channel.php]: probably better use findOrFail

File: app/Http/Controllers/API/BeatmapsController.php [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        // FIXME: scores are obtained via filename/checksum lookup for legacy reason (temporarily)
+        $filename = Request::input('f');
+        $checksum = Request::input('c');
+        $per_page = min(Request::input('n', 50), 50);
+        $page = max(Request::input('p', 1), 1);
+
+        $beatmap = Beatmap::where('filename', $filename)->where('checksum', $checksum)->first();
+
+        if (!$beatmap) {
+            return Response::json(['error' => 'not_found']);
```
Comment (nanaya) [inline app/Http/Controllers/API/BeatmapsController.php]: might as well `firstOrFail`? Also this one returns 200 on not found?

File: app/Http/Controllers/API/BeatmapsController.php [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        }
+
+        $beatmap_meta = fractal_api_serialize_item(
+            $beatmap,
+            new BeatmapTransformer()
+        );
+
+        $scores = $beatmap->scoresBest()->defaultScope()->offset(($page - 1) * $per_page)->limit($per_page);
```
Comment (nanaya) [inline app/Http/Controllers/API/BeatmapsController.php]: there's a `forPage()` method in query builder which does the same thing.

File: app/Http/Controllers/OAuthController.php (hunk 3 above)
Comment (nekodex) [inline app/Http/Controllers/OAuthController.php]: It's great this is mentioned in the Laravel controller documentation but not in the middleware documentation... 😌👍💯👍👌💯

File: app/Transformers/API/Chat/PrivateMessageTransformer.php [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+{
+    public function transform(PrivateMessage $message)
+    {
+        return [
+            'message_id' => $message->message_id,
+            'user_id' => $message->user_id,
+            'target_id' => $message->target_id,
+            'timestamp' => $message->timestamp->toDateTimeString(),
+            'content' => $message->content,
+            'sender' => $message->sender ? [
```
Comment (nanaya) [inline app/Transformers/API/Chat/PrivateMessageTransformer.php]: use `UserTransformer`?

File: app/Transformers/API/Chat/PrivateMessageTransformer.php [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+class PrivateMessageTransformer extends Fractal\TransformerAbstract
+{
+    public function transform(PrivateMessage $message)
+    {
+        return [
+            'message_id' => $message->message_id,
+            'user_id' => $message->user_id,
+            'target_id' => $message->target_id,
+            'timestamp' => $message->timestamp->toDateTimeString(),
```
Comment (nanaya) [inline app/Transformers/API/Chat/PrivateMessageTransformer.php]: use iso thingy?

File: config/oauth2.php [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+        ],
+        'password' => [
+            'class' => '\League\OAuth2\Server\Grant\PasswordGrant',
+            'callback' => '\App\Libraries\OAuth2PasswordGrantVerifier',
+            'access_token_ttl' => 3600,
+        ],
+        'refresh_token' => [
+            'class' => '\League\OAuth2\Server\Grant\RefreshTokenGrant',
+            'access_token_ttl' => 3600,
+            'refresh_token_ttl' => 7.884e+6, // ~3 months
```
Comment (nanaya) [inline config/oauth2.php]: oh noes what is this

Comment (nekodex) [inline config/oauth2.php]: https://en.wikipedia.org/wiki/Scientific_notation

File: app/Http/Controllers/API/ChatController.php [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+        );
+    }
+
+    public function privateMessages()
+    {
+        $since = intval(Request::input('since'));
+        $limit = intval(Request::input('limit', 50));
+
+        $messages = PrivateMessage::where('user_id', $this->current_user->user_id)
+            ->orWhere('target_id', $this->current_user->user_id)
```
Comment (nanaya) [inline app/Http/Controllers/API/ChatController.php]: looks wrong? or correct?

```
select * from `messages_private` where `user_id` = ? or `target_id` = ? and `message_id` > ?
                                       (           ) or (               and                )
```

https://dev.mysql.com/doc/refman/5.7/en/operator-precedence.html

File: config/oauth2.php (hunk 10 above)
Comment (nanaya) [inline config/oauth2.php]: (屮゜Д゜)屮

File: app/Models/Chat/PrivateMessage.php [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+use App\Models\User;
+
+class PrivateMessage extends Model
+{
+    protected $table = 'messages_private';
+    protected $primaryKey = 'message_id';
+    protected $dates = [
+        'timestamp',
+    ];
+    public $timestamps = false;
```
Comment (nanaya) [inline app/Models/Chat/PrivateMessage.php]: ensure this really is false? :best:

Comment (nekodex) [inline app/Models/Chat/PrivateMessage.php]: GOTTA BE SURE

**Pull request merged**